### PR TITLE
Incorporating XML schema changes

### PR DESCRIPTION
**Pull request by mazelife**
I've made some assumptions based on the data that may be correct or incorrect, and have attempted to document those assumptions where they appear in the code. In real life, these changes would probably warrant some discussion to properly flesh out the new requirements before updating any code. But in this case I focused on getting something working that seemed like it could be a reasonable interpretation of a database schema that properly maps to the XML data given. There are also some performance implications of the schema change (increased number of DB joins, for example) that would need to be carefully considered.

### CODE REVIEW

File: storage/models.py [hunk 1]
```diff
@@ -18,19 +18,43 @@ class Book(BaseModel):
...
+    @property
+    def title(self):
+        return self.latest_edition.title
+
+    @property
+    def description(self):
+        return self.latest_edition.description
+
+
+class Edition(BaseModel):
```
Comment (hendawy) [inline storage/models.py]: I believe adding an "Edition" model will make the model "Book" redundunt. In the database, it will be a table with one column, the id. Instead I suggest keeping the Book model the same but instead using an integer, auto-incremental, unique primary key "id" and have field "book_id" (currently is just book.id) that will be the identifier for the book, this case book_id will not be unique, it will be the same for all editions of the same book.

File: storage/tools.py [hunk 2]
```diff
@@ -2,25 +2,59 @@
...
-
-    book, created = Book.objects.get_or_create(pk=book_element.get('id'))
-    book.title = book_element.findtext('title')
-    book.description = book_element.findtext('description')
-
-    for alias in book_element.xpath('aliases/alias'):
-        scheme = alias.get('scheme')
-        value = alias.get('value')
-
+    book_id = book_element.get('id')
```
Comment (hendawy) [inline storage/tools.py]: What about logging in case the element doesn't have "id"?

File: storage/tools.py [hunk 3]
```diff
@@ -2,25 +2,59 @@
...
+    aliases = [(a.get('scheme'), a.get('value')) for a in book_element.xpath('aliases/alias')]
+    edition_version = book_element.findtext('version')
+    try:
+        edition_version = Decimal(edition_version)
+    except InvalidOperation:
+        raise BadDataFile("Invalid version data: {} is not a decimal number.".format(edition_version))
```
Comment (hendawy) [inline storage/tools.py]: Wouldn't it be helpful to log the element's id when raising an exception?

File: storage/tools.py [hunk 4]
```diff
@@ -2,25 +2,59 @@
...
+    except TypeError:  # Raised when there is no <version> element.
+        raise BadDataFile("The version number is missing from this file.")
+    try:
+        book = Book.objects.get(pk=book_id)
+    except Book.DoesNotExist:
+        book = None
+    # Try to match on aliases, all of which must agree.
+    books_matched = {}
+    for scheme, value in aliases:
+        for alias in Alias.objects.filter(scheme=scheme, value=value):
```
Comment (hendawy) [inline storage/tools.py]: If the number of aliases per document increase, its going to increase the number of the database read operations. What about something like: Alias.objects.filter(scheme__in=[item[0] for item in aliases], value__in=[item[1] for item in aliases])?

File: storage/tools.py [hunk 5]
```diff
@@ -2,25 +2,59 @@
...
+            if alias.book_id not in books_matched:
+                books_matched[alias.book_id] = alias.book
+            if len(books_matched) > 1:
+                raise BadDataFile("The aliases in this file match more than one book.")
```
Comment (hendawy) [inline storage/tools.py]: What about logging the failed aliase's schema and value for more information?

File: storage/tools.py [hunk 6]
```diff
@@ -2,25 +2,59 @@
...
+    # If a book was did not match by ID use the alias match if there was one, or create a new book.
+    if book is None:
+        if len(books_matched) == 1:
+            book = books_matched.values()[0]
```
Comment (hendawy) [inline storage/tools.py]: Why using only one aliase to identify a book? I think only ids should be used as an identifier.

File: storage/tools.py [hunk 7]
```diff
@@ -2,25 +2,59 @@
...
+        else:
+            book = Book.objects.create(pk=book_id)
+    # Handle create/update of the book's edition.
+    edition, created = Edition.objects.get_or_create(book_id=book.pk, version=edition_version)
+    edition.title = book_element.findtext('title')
+    edition.description = book_element.findtext('description')
+    edition.save()
+    # Handle create/update of the book's aliases.
+    for scheme, value in aliases:
         book.aliases.get_or_create(scheme=scheme, value=value)
```
Comment (hendawy) [inline storage/tools.py]: This will also create a lot of write operations to the database, what about using "bulk_create". Unless we anticipate using post_save or pre_save signals or extending the model's save method, I think it should be bulk_create.

File: storage/tests/test_tools.py [hunk 8]
```diff
@@ -18,6 +18,7 @@ def test_storage_tools_process_book_element_db(self):
         xml_str = '''
```
Comment (hendawy) [inline storage/tests/test_tools.py]: I believe that there a lot of boundary cases that are not covered after updating storage.tools.process_book_element.

File: storage/exceptions.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+class BadDataFile(Exception):
+    """
```
Comment (hendawy) [inline storage/exceptions.py]: The project's conviention is using triple single quoute for comment, not triple double quote
